### PR TITLE
Scheduling policies to retry and repeat effectful computations

### PR DESCRIPTION
**Pull request by truizlop**
## Related issues

Current implementation does not have a mechanism to retry or repeat effectful computations based on `IO` or `EnvIO`.

## Goal

Create a composable way of specifying retrial/repeat policies to re-run effectful computations. I used [ZIO](https://github.com/zio/zio/blob/fae35db0dd8c8fb89e52a76bdf2f22d63a81a4e1/core/shared/src/main/scala/zio/Schedule.scala) as an inspiration to achieve this.

## Implementation details

Functions to `retry` and `repeat` have been added to `EnvIO` and `IO` (the latter delegates on the former). These functions receive a `Schedule` that dictates if the effect needs to be reapplied or not.

Primitive `Schedules` have been added, where the most prominent ones are:

- `Schedule.recurs(n)`: recurs `n` times.
- `Schedule.forever()`: recurs forever.
- `Schedule.spaced(time)`: recurs forever, spacing each execution the specified time.
- `Schedule.linear(time)`: recurs forever, incrementing the delay between each attempt in a linear manner, using `time` as a baseline.
- `Schedule.exponential(time)`: recurs forever, incrementing the delay between each attempt in an exponential manner, using `time` as a baseline.
- `Schedule.fibonacci(time)`: recurs forever, incrementing the delay between each attempt following the Fibonacci sequence, using `time` as a baseline.
- `Schedule.duration(time)`: recurs until the specified `time` elapses.

These primitives can be composed in different manners, always resulting in new `Schedule`s that can be subsequently composed. The most prominent composition operators are:

- `s1.and(s2)`: executes as long as both schedules want to keep recursing, using the maximum of both delays.
- `s1.or(s2)`: executes as long as one of the schedules want to keep recursing, using the minimum of both delays.
- `s1.andThen(s2)`: executes `s1` until it states to finish, and then executes `s2`.

## Usage

An example of usage:

Retries the `io` operation 8 times, waiting 2 seconds between each attempt.

```swift
io.retry(Schedule.recurs(8).and(Schedule.spaced(.seconds(2)))
```

### CODE REVIEW

File: Sources/BowEffects/Data/Internal/DispatchTimeInterval+Extensions.swift [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import Foundation
+
+extension DispatchTimeInterval {
+    func toDouble() -> Double? {
+        switch self {
+        case .seconds(let value): return Double(value)
+        case .milliseconds(let value): return Double(value) * 0.001
+        case .microseconds(let value): return Double(value) * 0.000001
```
Comment (miguelangel-dev) [inline Sources/BowEffects/Data/Internal/DispatchTimeInterval+Extensions.swift]:
```suggestion
        case .microseconds(let value): return Double(value) * 0.000_001
```

I think it is easier to read

File: Sources/BowEffects/Data/Internal/DispatchTimeInterval+Extensions.swift [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+        case .nanoseconds(let value): return Double(value) * 0.000000001
```
Comment (miguelangel-dev) [inline Sources/BowEffects/Data/Internal/DispatchTimeInterval+Extensions.swift]:
```suggestion
        case .nanoseconds(let value): return Double(value) * 0.000_000_001
```

Comment (truizlop) [reply Sources/BowEffects/Data/Internal/DispatchTimeInterval+Extensions.swift]: Done!

File: Sources/BowEffects/Data/Schedule.swift [hunk 3]
```diff
@@ -0,0 +1,768 @@
...
+                    s2 <- that.update(self.extract(a, s.0), s.1),
+                    yield: (s1.get, s2.get))^
+            }
+        )
+    }
+    
+    /// Pipes the output of the provided schedule as the input of this schedule. Effects described by the provided schedule will always be performed before the effects described by this schedule.
+    ///
+    /// - Parameter that: Schedule to be piped before this one.
+    public func reversePipe<State2, C>(_ that: Schedule<R, C, State2, A>) -> Schedule<R, C, (State2, State), B> {
```
Comment (miguelangel-dev) [inline Sources/BowEffects/Data/Schedule.swift]: Maybe it is a dummy question, but I reverse like:
`public func reversePipe<State2, C>(_ that: Schedule<R, C, State2, B>) -> Schedule<R, C, (State2, State), A>`

Comment (truizlop) [reply Sources/BowEffects/Data/Schedule.swift]: No, this is like the operators `<<<` and `>>>`, with the following signatures:

`>>>`: `(f: (A) -> B, g: (B) -> C) -> (A) -> C`
`>>>`: `(g: (B) -> C, f: (A) -> B) -> (A) -> C`

The only thing you are reversing is the order of the parameters, not their type arguments.

File: Sources/BowEffects/Data/Schedule.swift [hunk 4]
```diff
@@ -0,0 +1,768 @@
...
+    /// - Parameter f: An effectful function that provides the delay time based on the output of the schedule.
+    public func addDelayM(_ f: @escaping (B) -> EnvIO<R, Never, TimeInterval>) -> Schedule<R, A, State, B> {
+        updated { upd in
+            { a, s in
+                let delay = EnvIO<R, Unit, TimeInterval>.var()
+                let s1 = EnvIO<R, Unit, State>.var()
+                return binding(
+                    delay <- f(self.extract(a, s)).unitError(),
+                    s1 <- upd(a, s),
+                    |<-EnvIO.sleep(delay.get),
```
Comment (miguelangel-dev) [inline Sources/BowEffects/Data/Schedule.swift]: smart delay...

File: Sources/BowEffects/Data/Schedule.swift (hunk 3 above)
Comment (miguelangel-dev) [reply Sources/BowEffects/Data/Schedule.swift]: Ah! Now I see the point, thanks

File: Sources/BowEffects/Data/Internal/DispatchTimeInterval+Extensions.swift (hunk 1 above)
Comment (miguelangel-dev) [reply Sources/BowEffects/Data/Internal/DispatchTimeInterval+Extensions.swift]: same here!

**Pull request merged**